homeassistant/components/telegram_bot/__init__.py

```diff
@@ -857,131 +858,102 @@ def leave_chat(self, chat_id=None):
-
-            data = {
-                ATTR_MSGID: msg_data["message_id"],
-                ATTR_CHAT_ID: msg_data["chat"]["id"],
-                ATTR_TEXT: msg_data["text"],
-            }
-            return True, data
-
-        _LOGGER.error("Incoming message does not have required data (%s)", msg_data)
-        return False, None
-
-    def process_message(self, data):
-        """Check for basic message rules and fire an event if message is ok."""
-        if ATTR_MSG in data or ATTR_EDITED_MSG in data:
-            event = EVENT_TELEGRAM_COMMAND
-            if ATTR_MSG in data:
-                data = data.get(ATTR_MSG)
-            else:
-                data = data.get(ATTR_EDITED_MSG)
-            message_ok, event_data = self._get_message_data(data)
-            if event_data is None:
-                return message_ok
-
-            if ATTR_MSGID in data:
-                event_data[ATTR_MSGID] = data[ATTR_MSGID]
-
-            if "text" in data:
-                if data["text"][0] == "/":
-                    pieces = data["text"].split(" ")
-                    event_data[ATTR_COMMAND] = pieces[0]
-                    event_data[ATTR_ARGS] = pieces[1:]
-                else:
-                    event_data[ATTR_TEXT] = data["text"]
-                    event = EVENT_TELEGRAM_TEXT
-            else:
-                _LOGGER.warning("Message without text data received: %s", data)
-                event_data[ATTR_TEXT] = str(data)
-                event = EVENT_TELEGRAM_TEXT
 
-            self.hass.bus.async_fire(event, event_data)
+    def handle_update(self, update: Update, context: CallbackContext):
+        """Handle updates from bot dispatcher set up by the respective platform."""
+        _LOGGER.debug("Handling update %s", update)
+        if not self.authorize_update(update):
+            return
+
+        # establish event type: text, command or callback_query
+        if update.callback_query:
+            # NOTE: Check for callback query first since effective message will be populated with the message
+            # in .callback_query (python-telegram-bot docs are wrong)
+            event_type, event_data = self._get_callback_query_event_data(
+                update.callback_query
+            )
+        elif update.effective_message:
+            event_type, event_data = self._get_message_event_data(
+                update.effective_message
+            )
+        else:
+            _LOGGER.warning("Unhandled update: %s", update)
             return True
-        if ATTR_CALLBACK_QUERY in data:
-            event = EVENT_TELEGRAM_CALLBACK
-            data = data.get(ATTR_CALLBACK_QUERY)
-            message_ok, event_data = self._get_message_data(data)
-            if event_data is None:
-                return message_ok
 
-            query_data = event_data[ATTR_DATA] = data[ATTR_DATA]
+        _LOGGER.debug("Firing event %s: %s", event_type, event_data)
+        self.hass.bus.async_fire(event_type, event_data)
```

Comment:
I don't think that `handle_update` is run in async context, ie inside the asyncio event loop. We need to use the sync api inside sync context.

Reply:
> I don't think that `handle_update` is run in async context, ie inside the asyncio event loop. We need to use the sync api inside sync context.

Thanks for the review @MartinHjelmare !
Switching to `bus.fire` works, but my test for the polling platform fails. I can see in the logs the event gets fired but I'm unable to catch it in a test the way I did earlier (`async_listen_once` with a `Mock` callback, and then `hass.async_block_till_done` + `assert_called_once` on mock). Do you have any suggestion for how to set up a fitting listener/mock?

Reply:
1. Patch the library to get access to the library callback. Eg make a pytest fixture that yields the patched mock.
2. Set up a listener to be able to assert fired events.
3. Set up the integration.
4. Schedule the callback on the executor thread pool with `hass.async_add_executor_job`, and wait for all tasks with `hass.async_block_till_done`.
5. Assert that the event was fired.